File: utils/Operators/UPDATE_CHEMISTRY_EQUILIBRIUM.m
function s = UPDATE_CHEMISTRY_EQUILIBRIUM(s, chemistry)
% UPDATE_CHEMISTRY_EQUILIBRIUM  Evaluate equilibrium thermochemical and transport properties.
%
%   s = UPDATE_CHEMISTRY_EQUILIBRIUM(s, chemistry)
%
%   Computes the equilibrium values of gamma_star, cv_star, dynamic
%   viscosity (mu), and thermal conductivity (k) from the chemistry lookup
%   tables. When chemical equilibrium is active, the s properties
%   are set to their equilibrium values; otherwise, only the _eq fields
%   are updated. For non-chemistry cases, uniform freestream values are
%   assigned. Derived quantities (Re_flow, Pr_flow) are also updated.
%
%   Inputs:
%       s  - Solution struct containing:
%                   .var.rho, .var.rho_u, .var.rho_v, .var.rho_E - Conserved variables
%                   .chemistry.is_chemistry_enabled - Logical, true if chemistry active
%                   .chemistry.chemical_equilibrium - Logical, true if in equilibrium mode
%                   .freestream.*            - Freestream reference values and
%                                              scaling factors
%       chemistry - Chemistry model struct with evaluation functions:
%                   .eval_gamma_star, .eval_cv_star, .eval_mu, .eval_k
%
%   Outputs:
%       s  - Updated s struct with fields:
%                   .var.gamma_star_eq, .var.gamma_star - Equilibrium ratio of specific heats
%                   .var.cv_star_eq, .var.cv_star       - Equilibrium specific heat at const. vol.
%                   .var.mu_star                        - Dynamic viscosity
%                   .var.Re_flow                        - Local Reynolds number field
%                   .var.k_star                         - Thermal conductivity
%                   .var.Pr_flow                        - Local Prandtl number field
%
%   Notes:
%       - Internal energy is computed as e = (rho_E - KE) / rho.
%       - Dimensional conversions use freestream scaling factors.
%
% Part of: Hypersonics Stability MATLAB Solver - Operators Module

    %% Compute internal energy from conserved variables
    rho = s.var.rho;
    e = (s.var.rho_E - (s.var.rho_u.^2 + s.var.rho_v.^2) ./ rho / 2) ./ rho;

    %% Gamma_star (ratio of specific heats)
    if s.chemistry.is_chemistry_enabled
        s.var.gamma_star_eq = chemistry.eval_gamma_star(s.freestream.rho_factor * rho, s.freestream.energy_factor * e);
        if s.chemistry.chemical_equilibrium
            s.var.gamma_star = s.var.gamma_star_eq;
        end
    else
        s.var.gamma_star_eq = s.freestream.gamma_star * ones(size(rho));
        s.var.gamma_star = s.var.gamma_star_eq;
    end

    %% cv_star (specific heat at constant volume)
    if s.chemistry.is_chemistry_enabled
        cv_star = chemistry.eval_cv_star(s.freestream.rho_factor * rho, s.freestream.energy_factor * e); % Dimensional specific heat at constant volume
        s.var.cv_star_eq = cv_star ./ s.freestream.cv;
        if s.chemistry.chemical_equilibrium
            s.var.cv_star = s.var.cv_star_eq;
        end
    else
        cv = s.freestream.cv * ones(size(rho)); % Dimensional specific heat at constant volume
        s.var.cv_star_eq = cv ./ cv;
        s.var.cv_star = s.var.cv_star_eq;
    end

    %% Dynamic viscosity (mu) and local Reynolds number
    if s.chemistry.is_chemistry_enabled
        s.var.mu_star = chemistry.eval_mu(s.freestream.rho_factor * rho, s.freestream.energy_factor * e) ./ s.freestream.mu;
        s.var.Re_flow = s.freestream.Re ./ s.var.mu_star;
    else
        s.var.mu_star = s.freestream.mu ./ s.freestream.mu * ones(size(rho));
        s.var.Re_flow = s.freestream.Re * ones(size(rho));
    end

    %% Thermal conductivity (k) and local Prandtl number
    if s.chemistry.is_chemistry_enabled
        s.var.k_star = chemistry.eval_k(s.freestream.rho_factor * rho, s.freestream.energy_factor * e) ./ s.freestream.k;
        s.var.Pr_flow =  s.var.mu_star .* s.freestream.mu .* s.var.cv_star .* s.freestream.cv .* s.var.gamma_star ./ s.var.k_star / s.freestream.k;
    else
        s.var.k_star = s.freestream.k / s.freestream.k * ones(size(rho));
        s.var.Pr_flow = s.freestream.Pr * ones(size(rho));
    end
end

File: utils/Operators/UPDATE_CHEMISTRY_EQUILIBRIUM.jl
function UPDATE_CHEMISTRY_EQUILIBRIUM(s::Dict{String, Any}, chemistry::Dict{String, Any})
# UPDATE_CHEMISTRY_EQUILIBRIUM - Evaluate equilibrium thermochemical and transport properties.
#
# Computes the equilibrium values of gamma_star, cv_star, dynamic
# viscosity (mu), and thermal conductivity (k) from the chemistry lookup
# tables. Derived quantities (Re_flow, Pr_flow) are also updated.
#
# Part of: Hypersonics Stability Julia Solver - Operators Module

    ## Extract dict refs
    var  = s["var"]::Dict{String, Any}
    chem = s["chemistry"]::Dict{String, Any}
    fs   = s["freestream"]::Dict{String, Any}

    ## Compute internal energy from conserved variables
    rho       = var["rho"]::Matrix{Float64}
    rho_E     = var["rho_E"]::Matrix{Float64}
    rho_u_arr = var["rho_u"]::Matrix{Float64}
    rho_v_arr = var["rho_v"]::Matrix{Float64}
    e = @. (rho_E - (rho_u_arr^2 + rho_v_arr^2) / rho / 2) / rho

    rho_factor    = fs["rho_factor"]::Float64
    energy_factor = fs["energy_factor"]::Float64
    is_chem       = chem["is_chemistry_enabled"]::Bool
    is_eq         = chem["chemical_equilibrium"]::Bool

    ## Pre-compute dimensional state once (avoids 4x redundant allocations)
    if is_chem
        rho_dim = @. rho_factor * rho
        e_dim   = @. energy_factor * e
    end

    ## Gamma_star (ratio of specific heats)
    if is_chem
        eval_gamma_star = chemistry["eval_gamma_star"]
        var["gamma_star_eq"] = eval_gamma_star(rho_dim, e_dim)
        if is_eq
            var["gamma_star"] = var["gamma_star_eq"]
        end
    else
        gs_val = fs["gamma_star"]::Float64
        gamma_star_arr = var["gamma_star"]::Matrix{Float64}
        fill!(gamma_star_arr, gs_val)
        var["gamma_star_eq"] = gamma_star_arr
    end

    ## cv_star (specific heat at constant volume)
    cv_fs = fs["cv"]::Float64
    if is_chem
        eval_cv_star = chemistry["eval_cv_star"]
        cv_star_dim = eval_cv_star(rho_dim, e_dim)
        var["cv_star_eq"] = @. cv_star_dim / cv_fs
        if is_eq
            var["cv_star"] = var["cv_star_eq"]
        end
    else
        cv_star_arr = var["cv_star"]::Matrix{Float64}
        fill!(cv_star_arr, 1.0)
        var["cv_star_eq"] = cv_star_arr
    end

    ## Dynamic viscosity (mu) and local Reynolds number
    Re_fs = Float64(fs["Re"])
    mu_fs = fs["mu"]::Float64
    if is_chem
        eval_mu = chemistry["eval_mu"]
        var["mu_star"] = @. eval_mu(rho_dim, e_dim) / mu_fs
        mu_star_arr = var["mu_star"]::Matrix{Float64}
        var["Re_flow"] = @. Re_fs / mu_star_arr
    else
        mu_star_arr = var["mu_star"]::Matrix{Float64}
        fill!(mu_star_arr, 1.0)
        Re_flow_arr = var["Re_flow"]::Matrix{Float64}
        fill!(Re_flow_arr, Re_fs)
    end

    ## Thermal conductivity (k) and local Prandtl number
    k_fs = fs["k"]::Float64
    if is_chem
        eval_k = chemistry["eval_k"]
        var["k_star"] = @. eval_k(rho_dim, e_dim) / k_fs
        gamma_star_arr = var["gamma_star"]::Matrix{Float64}
        cv_star_arr    = var["cv_star"]::Matrix{Float64}
        mu_star_arr    = var["mu_star"]::Matrix{Float64}
        k_star_arr     = var["k_star"]::Matrix{Float64}
        var["Pr_flow"] = @. mu_star_arr * mu_fs * cv_star_arr * cv_fs * gamma_star_arr / k_star_arr / k_fs
    else
        Pr_fs = fs["Pr"]::Float64
        k_star_arr = var["k_star"]::Matrix{Float64}
        fill!(k_star_arr, 1.0)
        Pr_flow_arr = var["Pr_flow"]::Matrix{Float64}
        fill!(Pr_flow_arr, Pr_fs)
    end

    return s
end
